### PR TITLE
feature: configurable retry params for clouds

### PR DESCRIPTION
**Pull request by rahulreddy**
This introduces configurable retry parameters for all classes that use
BackbeatTask for handling retry mechanism. For now, for Zenko, retry
tunables are available per-cloud which affect how many times an entry
is retried during temporary failures.

### CODE REVIEW

**Comment by rahulreddy** [comment]
@bert-e wait

File: extensions/replication/tasks/ReplicateObject.js [hunk 1]
```diff
@@ -34,10 +34,14 @@ class ReplicateObject extends BackbeatTask {
     constructor(qp) {
         const qpState = qp.getStateVars();
         super({
-            retryTimeoutS: qpState.repConfig.queueProcessor.retryTimeoutS,
+            retry: qpState.repConfig.queueProcessor.retry,
         });
         Object.assign(this, qpState);
-
+        this.destType = this.destConfig.bootstrapList
+        .find(endpoint => endpoint.site === this.site);
```
Comment (bennettbuchanan) [inline extensions/replication/tasks/ReplicateObject.js]: After finding the site object: e.g.

```
{ "site": "aws-location", "type": "aws_s3" }
```

It looks like we need to also get the `type` from that object.

File: extensions/replication/tasks/ReplicateObject.js [hunk 2]
```diff
@@ -34,10 +34,14 @@ class ReplicateObject extends BackbeatTask {
...
+        const retryParams = this.repConfig.queueProcessor.retry[this.destType];
```
Comment (bennettbuchanan) [inline extensions/replication/tasks/ReplicateObject.js]: We'll have to have a way of accounting for a bootstrap list type of `aws_s3` and the newly added retry config property `aws`.

File: conf/config.json [hunk 3]
```diff
@@ -97,7 +97,48 @@
...
+                        "maxRetries": 5,
+                        "timeoutS": 300,
+                        "backoff": {
+                            "min": 1000,
+                            "max": 30000,
+                            "jitter": 0.1,
+                            "factor": 1.5
+                        }
+                    },
+                    "default": {
```
Comment (bennettbuchanan) [inline conf/config.json]: Looks like this will also apply to a Scality RING destination because there is no `site` property. Thinking aloud here: maybe we could move the default values to the Joi config validator (i.e. using the `default()` method) and then rename this something like "scality".

File: lib/tasks/BackbeatTask.js [hunk 4]
```diff
@@ -26,10 +24,9 @@ class BackbeatTask {
     retry(args, done) {
         const { actionDesc, logFields,
                 actionFunc, shouldRetryFunc, onRetryFunc, log } = args;
-        const backoffCtx = new BackOff(BACKOFF_PARAMS);
+        const backoffCtx = new BackOff(this.retryParams.backoff);
```
Comment (bennettbuchanan) [inline lib/tasks/BackbeatTask.js:23]: 👍

File: extensions/replication/tasks/ReplicateObject.js (hunk 1 above)
Comment (rahulreddy) [reply extensions/replication/tasks/ReplicateObject.js]: Ah you are right. I should be pulling `type` instead of site.

File: extensions/replication/tasks/ReplicateObject.js (hunk 2 above)
Comment (rahulreddy) [reply extensions/replication/tasks/ReplicateObject.js]: Yeah it should be called `aws_s3`. I need to find a way to have one source of truth for types (refactor for later).

File: conf/config.json (hunk 3 above)
Comment (rahulreddy) [reply conf/config.json]: Yeah, that's a good idea. I will make the change.

File: conf/config.json [hunk 5]
```diff
@@ -97,12 +105,61 @@
             "monitorReplicationFailureExpiryTimeS": 86400,
             "queueProcessor": {
                 "groupId": "backbeat-replication-group",
-                "retryTimeoutS": 300,
+                "retry": {
+                    "aws": {
+                        "maxRetries": 5,
+                        "timeoutS": 300,
+                        "backoff": {
+                            "min": 1000,
```
Comment (jonathan-gramain) [inline conf/config.json]: Didn't you want to take the opportunity to adjust the default values for all clouds, e.g. setting a larger min backoff and larger timeout by default? Not perfect for all environments of course but hopefully closer to reasonable defaults in production env.

File: lib/tasks/BackbeatTask.js [hunk 6]
```diff
@@ -1,35 +1,16 @@
...
-    constructor(config) {
-        this.config = joi.attempt(config || {}, backbeatTaskConfigJoi,
-                                  'invalid backbeat task config');
-    }
 
     retry(args, done) {
         const { actionDesc, logFields,
                 actionFunc, shouldRetryFunc, onRetryFunc, log } = args;
-        const backoffCtx = new BackOff(BACKOFF_PARAMS);
+        const backoffCtx = new BackOff(this.retryParams.backoff);
```
Comment (jonathan-gramain) [inline lib/tasks/BackbeatTask.js:23]: Since we don't provide the retry params in the constructor anymore, I would rather populate them with default values in the constructor, and provide a setter to change them. That would avoid taking the risk of forgetting to set the retry params and raising an exception.

File: lib/tasks/BackbeatTask.js [hunk 7]
```diff
@@ -42,30 +23,32 @@ class BackbeatTask {
...
 
                 const now = Date.now();
-                if (now > (startTime + self.config.retryTimeoutS * 1000)) {
-                    log.error(
-                        `retried for too long to ${actionDesc}, giving up`,
+                const retriesExceeded = nbRetries > this.retryParams.maxRetries;
+                const timeoutReached = now >
+                    (startTime + this.retryParams.timeoutS * 1000);
+                if (retriesExceeded || timeoutReached) {
+                    log.error('giving up replication as retries exceeded',
```
Comment (jonathan-gramain) [inline lib/tasks/BackbeatTask.js]: `actionDesc` should be used for the log instead of "replication", also I would rephrase "retries exceeded" to "giving up retries" to be agnostic as to why we stopped retrying (the reason can be found as `retriesExceeded` or `timeoutReached` in the log params).

File: lib/tasks/BackbeatTask.js [hunk 8]
```diff
@@ -42,30 +23,32 @@ class BackbeatTask {
...
             }
             if (nbRetries > 0) {
-                const now = Date.now();
-                log.info(`succeeded to ${actionDesc} after retries`,
-                         Object.assign({
-                             nbRetries,
-                             retryTotalMs: `${now - startTime}`,
-                         }, logFields || {}));
+                const retryTotalMs = Date.now() - startTime;
+                log.info('replication succeeded after retries',
```
Comment (jonathan-gramain) [inline lib/tasks/BackbeatTask.js]: Same thing, use `actionDesc`.

**Review by jonathan-gramain — COMMENTED** [review]
Some early comments

**Comment by jonathan-gramain** [comment]
I was thinking also as a general idea, could it be better to consider stopping retries after *both* timeout *and* maximum number of retries are exceeded, instead of *either* one? I'm considering the case of putting a very large object, where it can be useful to retry, say at least once, but the timeout will surely be exceeded after the first failure if not very large. As an alternative we could also provide a minimum number of retries.

File: conf/config.json (hunk 5 above)
Comment (rahulreddy) [reply conf/config.json]: Yeah this PR is a bit premature at the moment. I will be apply a default of 15 min for all clouds. Since we don't have any real world data at the moment we can live with that and the tunable should help as it can be changed from Zenko.

File: lib/tasks/BackbeatTask.js (hunk 7 above)
Comment (rahulreddy) [reply lib/tasks/BackbeatTask.js]: I will make it more generic as it will be easier to filter and search for that specific `actionDesc` in Kibana.

**Comment by rahulreddy** [comment]
I think having both timeout and minimum number of retries would have negative effect on small objects esp. 0-byte ones or MD updates only entries. I like the idea of doing min. retries. but it's confusing as to which one trumps the other (#of retries or timeout). It's easier to think of the logic the other way around i.e `either or`, it would mean we would retry the object upto the maxRetries param in the case of small objects and we would try at least timeoutS (this needs to be adjusted to factor in large objects).

**Comment by jonathan-gramain** [comment]
Actually, for 0 byte object, or any small object, it seems to me desirable to satisfy at least the maximum retry time specified, to guarantee we can replicate everything if a cloud is down for some time less than this maximum, absorbing transient unavailability (may be due to the cloud but also to the local client network as well).

On the other hand, large objects may still be retried for longer, because they may have been hit by a random connectivity issue in the middle of a large transfer. In either case, especially for small objects, the total number of retries will be kept low by the exponential backoff (so roughly `nretries = O(log(retryTimeout/minRetryDelay))` which should be relatively low, ignoring the `maxRetryDelay` param).

**Comment by rahulreddy** [comment]
> Actually, for 0 byte object, or any small object, it seems to me desirable to satisfy at least the maximum retry time specified, to guarantee we can replicate everything if a cloud is down for some time less than this maximum, absorbing transient unavailability (may be due to the cloud but also to the local client network as well).

I don't think factoring issues like cloud downtime is feasible, that would be handled by the manual retrying of failed objects feature. This is to absorb interim nuances like AWS giving 404s even when bucket exists, socket hangups etc. At the moment we are either retrying too quickly or failing too quickly, that's what we are trying to address in this PR.

File: extensions/replication/tasks/ReplicateObject.js [hunk 9]
```diff
@@ -33,11 +33,19 @@ class ReplicateObject extends BackbeatTask {
...
-            retryTimeoutS: qpState.repConfig.queueProcessor.retryTimeoutS,
-        });
+        super();
         Object.assign(this, qpState);
-
+        this.retryParams = this.repConfig.queueProcessor.retry.scality;
+        if (this.destConfig && this.destConfig.bootstrapList) {
+            const destination = this.destConfig.bootstrapList
+                .find(endpoint => endpoint.site === this.site) || {};
+            this.destType = destination.type;
```
Comment (philipyoo) [inline extensions/replication/tasks/ReplicateObject.js:42]: I think we can remove this?

Comment (rahulreddy) [reply extensions/replication/tasks/ReplicateObject.js:42]: I wanted to change this https://github.com/scality/backbeat/blob/development/8.1/extensions/replication/tasks/MultipleBackendTask.js#L22 as this is being evaluated on every request.
I will follow up with another PR for that.

**Review by philipyoo — APPROVED** [review]
lgtm 👍

File: conf/config.json [hunk 10]
```diff
@@ -97,12 +106,62 @@
             "monitorReplicationFailureExpiryTimeS": 86400,
             "queueProcessor": {
                 "groupId": "backbeat-replication-group",
-                "retryTimeoutS": 300,
+                "retry": {
+                    "aws_s3": {
+                        "maxRetries": 5,
+                        "timeoutS": 900,
```
Comment (bennettbuchanan) [inline conf/config.json:107]: This increase looks good to me. Just want to confirm that it's intended to increase the default timeout for public cloud locations.

File: lib/tasks/BackbeatTask.js [hunk 11]
```diff
@@ -1,37 +1,30 @@
...
-    constructor(config) {
-        this.config = joi.attempt(config || {}, backbeatTaskConfigJoi,
-                                  'invalid backbeat task config');
+
+    constructor() {
+        this.retryParams = {
+            timeoutS: 300,
+            backoff: {
+                min: 1000,
+                max: 300000,
```
Comment (bennettbuchanan) [inline lib/tasks/BackbeatTask.js:13]: ~300000~ 30000

File: lib/config/configItems.joi.js [hunk 12]
```diff
@@ -51,10 +51,22 @@ const zenkoAuthJoi = joi.object({
     account: joi.string().required(),
 });
 
+const retryParamsJoi = joi.object({
```
Comment (bennettbuchanan) [inline lib/config/configItems.joi.js:54]: Seems like these default values are a combination of scality and public cloud destinations. Also, wondering why `0.2` jitter, whereas elsewhere it is consistently `0.1`.

File: lib/tasks/BackbeatTask.js [hunk 13]
```diff
@@ -42,33 +35,40 @@ class BackbeatTask {
...
-                    log.error(
-                        `retried for too long to ${actionDesc}, giving up`,
+                const retriesMaxedOut = nbRetries >=
+                    this.retryParams.maxRetries;
+                const timeoutReached = now >=
+                    (startTime + this.retryParams.timeoutS * 1000);
+                // Give up if max retries reached or if timeout reached and
+                // the entry has been retried at least once
+                if (retriesMaxedOut || (timeoutReached && nbRetries > 0)) {
+                    log.error('giving up processing as retries are maxed out',
```
Comment (bennettbuchanan) [inline lib/tasks/BackbeatTask.js]: Minor: could be useful to make the message conditional whether the timeout or number of retries was reached.

**Review by bennettbuchanan — APPROVED** [review]
I've left some minor comments around config value consistency and logging. Otherwise, LGTM! 🚢

File: conf/config.json (hunk 10 above)
Comment (rahulreddy) [reply conf/config.json:107]: That's correct, 15 min. or 5 retries should be a good start. We will refine this as we know better.

File: lib/tasks/BackbeatTask.js (hunk 11 above)
Comment (rahulreddy) [reply lib/tasks/BackbeatTask.js:13]: It's 300K ms. as in 300S which is the timeout. Isn't that correct?

File: lib/tasks/BackbeatTask.js (hunk 13 above)
Comment (rahulreddy) [reply lib/tasks/BackbeatTask.js]: I added that as log properties so that the message can be used for searching consistently and filtering the log properties `retriesMaxedOut` or `timeoutReached` would give the info where it timed out or had max retries reached.

File: lib/tasks/BackbeatTask.js (hunk 11 above)
Comment (bennettbuchanan) [reply lib/tasks/BackbeatTask.js:13]: I see, it looks like the other values set at `30000` need to be updated in that case.

File: lib/config/configItems.joi.js (hunk 12 above)
Comment (rahulreddy) [reply lib/config/configItems.joi.js:54]: It's a remnant of changes that did not propagate. I will amend this, `0.2` or `0.1` has medium impact, so just sticking to one for consistency.

File: lib/tasks/BackbeatTask.js (hunk 13 above)
Comment (bennettbuchanan) [reply lib/tasks/BackbeatTask.js]: Good point re: searching logs.

File: extensions/replication/queueProcessor/QueueProcessor.js [hunk 14]
```diff
@@ -58,7 +58,7 @@ class QueueProcessor extends EventEmitter {
      *   specific to queue processor
      * @param {String} repConfig.queueProcessor.groupId - kafka
      *   consumer group ID
-     * @param {String} repConfig.queueProcessor.retryTimeoutS -
+     * @param {Object} repConfig.queueProcessor.retry -
      *   number of seconds before giving up retries of an entry
```
Comment (jonathan-gramain) [inline extensions/replication/queueProcessor/QueueProcessor.js]: The comment is off now

File: lib/tasks/BackbeatTask.js (hunk 11 above)
Comment (rahulreddy) [reply lib/tasks/BackbeatTask.js:13]: I just saw that, thanks for pointing out. I updated the values correctly to be 300K

File: lib/tasks/BackbeatTask.js (hunk 13 above)
Comment (jonathan-gramain) [inline lib/tasks/BackbeatTask.js]: "retries are maxed out" may be understood as "number of retries maxed out", while it may also have timed out

**Review by jonathan-gramain — APPROVED** [review]

**Comment by rahulreddy** [comment]
@bert-e reset
